File: scripts/sources/population.ts
/**
 * Build population database from Census 2021 TS001
 *
 * Usage: bun run scripts/sources/population.ts
 */

/* biome-ignore-all lint/suspicious/noConsole: Build script uses console for progress */

import { Database } from 'bun:sqlite';
import { mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { Glob } from 'bun';
import { createBatchInserter, downloadFile, extractZip, findColumnIndex, parseCsvLine, SOURCES_DIR, toInt, withTempDir } from '../utils.ts';

/**
 * Source page: https://www.nomisweb.co.uk/sources/census_2021_bulk
 * Direct download: https://www.nomisweb.co.uk/output/census/2021/census2021-ts001.zip
 */
const TS001_ZIP_URL = 'https://www.nomisweb.co.uk/output/census/2021/census2021-ts001.zip';
const DB_PATH = join(SOURCES_DIR, 'population.db');
mkdirSync(SOURCES_DIR, { recursive: true });

async function findLsoaCsv(extractDir: string): Promise<string> {
	const glob = new Glob('**/census2021-ts001-lsoa.csv');
	const files = Array.from(glob.scanSync(extractDir));
	if (files.length === 0) throw new Error('LSOA CSV not found in TS001 archive');
	return join(extractDir, files[0] ?? '');
}

async function buildDatabase(csvPath: string): Promise<void> {
	console.log('Building population database...');
	console.log(`Source: ${csvPath}`);
	console.log(`Output: ${DB_PATH}\n`);

	const content = await Bun.file(csvPath).text();
	const lines = content.split(/\r?\n/).filter(Boolean);
	if (lines.length === 0) throw new Error('Empty TS001 CSV');

	const header = parseCsvLine(lines[0] ?? '');
	const geoCodeIdx = findColumnIndex(header, ['geography code']);
	const totalIdx = findColumnIndex(header, ['all usual residents', 'total']);

	if (geoCodeIdx < 0 || totalIdx < 0) {
		throw new Error('Required TS001 columns not found');
	}

	const db = new Database(DB_PATH, { create: true });
	db.exec(`
		DROP TABLE IF EXISTS population;
		CREATE TABLE population (
			lsoa_code TEXT PRIMARY KEY,
			population INTEGER
		);
	`);

	const inserter = createBatchInserter<[string, number | null]>(db, 'INSERT INTO population (lsoa_code, population) VALUES (?, ?)', { batchSize: 5000 });

	for (let i = 1; i < lines.length; i++) {
		const cols = parseCsvLine(lines[i] ?? '');
		const lsoa = cols[geoCodeIdx]?.trim();
		if (!lsoa) continue;
		inserter.add([lsoa, toInt(cols[totalIdx])]);
	}

	const totalRows = inserter.flush();
	console.log(`Inserted ${totalRows.toLocaleString()} LSOAs.`);
	db.close();
}

export async function build(): Promise<void> {
	await withTempDir('population', async (tempDir) => {
		const zipPath = join(tempDir, 'ts001.zip');
		console.log('Downloading Census TS001...');
		console.log(`URL: ${TS001_ZIP_URL}\n`);
		await downloadFile(TS001_ZIP_URL, zipPath);
		console.log('Download complete.\n');

		const extractDir = join(tempDir, 'extract');
		console.log('Extracting ZIP archive...');
		await extractZip(zipPath, extractDir);
		console.log('Extraction complete.\n');

		const csvPath = await findLsoaCsv(extractDir);
		await buildDatabase(csvPath);
	});
}

if (import.meta.main) {
	build().catch((err) => {
		console.error('Population build failed:', err);
		process.exit(1);
	});
}

File: crates/let-sdk/src/sources/population.rs
#![forbid(unsafe_code)]

use std::path::Path;
use std::{env, path::PathBuf};

use csv::{ReaderBuilder, StringRecord};

use crate::errors::{ErrorCode, LetError, Result};

use super::common::{
    download_file, extract_zip, find_column_index, find_first_matching_file, open_source_db,
    to_i64, with_temp_dir,
};

const TS001_URL: &str = "https://www.nomisweb.co.uk/output/census/2021/census2021-ts001.zip";

pub fn build(db_path: &Path) -> Result<usize> {
    let temp = with_temp_dir()?;
    let zip_path = resolve_input_zip_path(&temp)?;
    let extract_dir = temp.path().join("extract");

    extract_zip(&zip_path, &extract_dir)?;

    let csv_path = find_first_matching_file(&extract_dir, &|path| {
        path.file_name()
            .and_then(|value| value.to_str())
            .is_some_and(|name| name.eq_ignore_ascii_case("census2021-ts001-lsoa.csv"))
    })?
    .ok_or_else(|| {
        LetError::new(
            ErrorCode::NotFound,
            "LSOA CSV not found in census TS001 archive".to_owned(),
            "verify population source archive contents",
        )
    })?;

    let mut reader = ReaderBuilder::new()
        .has_headers(true)
        .flexible(true)
        .from_path(&csv_path)
        .map_err(|error| {
            LetError::new(
                ErrorCode::Parse,
                format!("failed to open population CSV: {error}"),
                "verify population source file format",
            )
        })?;

    let headers = reader
        .headers()
        .map_err(|error| {
            LetError::new(
                ErrorCode::Parse,
                format!("failed to read population CSV headers: {error}"),
                "verify population source file format",
            )
        })?
        .iter()
        .map(|value| value.trim().to_lowercase())
        .collect::<Vec<_>>();

    let geo_code = find_column_index(&headers, &["geography code"]).ok_or_else(|| {
        LetError::new(
            ErrorCode::Parse,
            "missing geography code column in population CSV".to_owned(),
            "verify population source file format",
        )
    })?;

    let total =
        find_column_index(&headers, &["all usual residents", "total"]).ok_or_else(|| {
            LetError::new(
                ErrorCode::Parse,
                "missing total population column in population CSV".to_owned(),
                "verify population source file format",
            )
        })?;

    let mut connection = open_source_db(db_path)?;
    connection.execute_batch(
        "
        DROP TABLE IF EXISTS population;
        CREATE TABLE population (
            lsoa_code TEXT PRIMARY KEY,
            population INTEGER
        );
        ",
    )?;

    let tx = connection.transaction()?;
    let mut statement =
        tx.prepare("INSERT OR REPLACE INTO population (lsoa_code, population) VALUES (?1, ?2)")?;

    let mut inserted = 0usize;
    for row in reader.records() {
        let row = row.map_err(csv_err)?;
        let Some(lsoa_code) = cell(&row, geo_code) else {
            continue;
        };

        statement.execute(rusqlite::params![lsoa_code, to_i64(cell(&row, total))])?;
        inserted += 1;
    }

    drop(statement);
    tx.commit()?;
    connection.execute_batch("VACUUM; ANALYZE;")?;

    Ok(inserted)
}

fn resolve_input_zip_path(temp: &tempfile::TempDir) -> Result<PathBuf> {
    if let Ok(local_path) = env::var("POPULATION_TS001_ZIP_PATH") {
        return Ok(PathBuf::from(local_path));
    }

    let zip_path = temp.path().join("population-ts001.zip");
    let url = env::var("POPULATION_TS001_ZIP_URL").unwrap_or_else(|_| TS001_URL.to_owned());
    download_file(&url, &zip_path)?;
    Ok(zip_path)
}

fn cell(row: &StringRecord, idx: usize) -> Option<&str> {
    row.get(idx)
        .map(str::trim)
        .and_then(|value| (!value.is_empty()).then_some(value))
}

fn csv_err(error: csv::Error) -> LetError {
    LetError::new(
        ErrorCode::Parse,
        format!("failed to parse population CSV row: {error}"),
        "verify population source file format",
    )
}
